strider/query_planner.py

```diff
@@ -169,179 +182,116 @@ async def qg_to_og(
+    for edge in graph['edges'].values():
+        sub = graph['nodes'][edge['subject']]
+        obj = graph['nodes'][edge['object']]
 
         for op in operation_kp_map:
+            # Check if forward operation matches
             if (
-                    op.source_category in source["category"]
-                    and op.target_category in target["category"]
-                    and op.edge_predicate in edge["predicate"]
+                    op.source_category in sub["category"]
+                    and op.target_category in obj["category"]
+                    and op.edge_predicate in [f"-{p}->" for p in edge["predicate"]]
             ):
-                source['filtered_categories'].add(op.source_category)
-                target['filtered_categories'].add(op.target_category)
-                edge['filtered_predicates'].add(op.edge_predicate)
-
-    # Filter down node categories using those
-    # we have recieved from the KP registry
-    # to limit the number of permutations.
+                sub['filtered_categories'].add(op.source_category)
+                obj['filtered_categories'].add(op.target_category)
+                edge['filtered_predicates'].add(op.edge_predicate[1:-2])
+            # Reverse
+            if (
+                    op.source_category in obj["category"]
+                    and op.target_category in sub["category"]
+                    and op.edge_predicate in [f"<-{p}-" for p in edge["predicate"]]
+            ):
+                obj['filtered_categories'].add(op.source_category)
+                sub['filtered_categories'].add(op.target_category)
+                edge['filtered_predicates'].add(op.edge_predicate[2:-1])
 
     # We only filter nodes that are touching at least one edge ("connected")
     connected_nodes = set()
-    for edge in operation_graph['edges'].values():
-        connected_nodes.add(edge['source'])
-        connected_nodes.add(edge['target'])
+    for edge in graph['edges'].values():
+        connected_nodes.add(edge['subject'])
+        connected_nodes.add(edge['object'])
     for node_id in connected_nodes:
-        node = operation_graph['nodes'][node_id]
+        node = graph['nodes'][node_id]
         if 'category' not in node:
             continue
         node['category'] = list(node.pop('filtered_categories'))
 
     # Also filter edge predicates as well
-    for edge in operation_graph['edges'].values():
+    for edge in graph['edges'].values():
         edge['predicate'] = list(edge.pop('filtered_predicates'))
     return None
 
 
-# pylint: disable=too-many-locals
-async def find_valid_permutations(
-        operation_graph,
-        kp_registry: Registry = None,
-        logger: logging.Logger = LOGGER,
-) -> list[dict]:
-    """
-    Given an operation graph, generate a list of operation graphs
-    that are solvable, and annotate those with a 'request_kp' property
-    that shows which KPs to contact to get results.
+def get_next_nodes(graph, path):
```

Comment:
Type hints might help here, especially because this graph format is not the typical one we deal with.